File: src/client/acontext-py/src/acontext/types/session.py
"""Type definitions for session, message, and task resources."""

from typing import Any, Literal, NotRequired, TypedDict, Union

from pydantic import BaseModel, Field


class RemoveToolResultParams(TypedDict, total=False):
    """Parameters for the remove_tool_result edit strategy.

    Attributes:
        keep_recent_n_tool_results: Number of most recent tool results to keep with original content.
            Defaults to 3 if not specified.
        tool_result_placeholder: Custom text to replace old tool results with.
            Defaults to "Done" if not specified.
        keep_tools: List of tool names that should never have their results removed.
            Tool results from these tools are always kept regardless of keep_recent_n_tool_results.
        gt_token: Only remove tool results whose text has more than this many tokens.
            If omitted, all tool results are eligible for removal.
    """

    keep_recent_n_tool_results: NotRequired[int]
    tool_result_placeholder: NotRequired[str]
    keep_tools: NotRequired[list[str]]
    gt_token: NotRequired[int]


class RemoveToolResultStrategy(TypedDict):
    """Edit strategy to replace old tool results with placeholder text.

    Example:
        {"type": "remove_tool_result", "params": {"keep_recent_n_tool_results": 5, "tool_result_placeholder": "Cleared"}}
    """

    type: Literal["remove_tool_result"]
    params: RemoveToolResultParams


class RemoveToolCallParamsParams(TypedDict, total=False):
    """Parameters for the remove_tool_call_params edit strategy.

    Attributes:
        keep_recent_n_tool_calls: Number of most recent tool calls to keep with full parameters.
            Defaults to 3 if not specified.
        keep_tools: List of tool names that should never have their parameters removed.
            Tool calls for these tools always keep their full parameters regardless of keep_recent_n_tool_calls.
        gt_token: Only remove tool call params whose arguments have more than this many tokens.
            If omitted, all tool calls are eligible for removal.
    """

    keep_recent_n_tool_calls: NotRequired[int]
    keep_tools: NotRequired[list[str]]
    gt_token: NotRequired[int]


class RemoveToolCallParamsStrategy(TypedDict):
    """Edit strategy to remove parameters from old tool-call parts.

    Keeps the most recent N tool calls with full parameters, replacing older
    tool call arguments with empty JSON "{}". The tool call ID and name remain
    intact so tool-results can still reference them.

    Example:
        {"type": "remove_tool_call_params", "params": {"keep_recent_n_tool_calls": 5}}
    """

    type: Literal["remove_tool_call_params"]
    params: RemoveToolCallParamsParams


class TokenLimitParams(TypedDict):
    """Parameters for the token_limit edit strategy.

    Attributes:
        limit_tokens: Maximum number of tokens to keep. Required parameter.
            Messages will be removed from oldest to newest until total tokens <= limit_tokens.
            Tool-call and tool-result pairs are always removed together.
    """

    limit_tokens: int


class TokenLimitStrategy(TypedDict):
    """Edit strategy to truncate messages based on token count.

    Removes oldest messages until the total token count is within the specified limit.
    Maintains tool-call/tool-result pairing - when removing a message with tool-calls,
    the corresponding tool-result messages are also removed.

    Example:
        {"type": "token_limit", "params": {"limit_tokens": 20000}}
    """

    type: Literal["token_limit"]
    params: TokenLimitParams


class MiddleOutParams(TypedDict):
    """Parameters for the middle_out edit strategy.

    Attributes:
        token_reduce_to: Target token count to reduce the prompt to. Required parameter.
    """

    token_reduce_to: int


class MiddleOutStrategy(TypedDict):
    """Edit strategy to reduce prompt size by removing middle messages.

    Example:
        {"type": "middle_out", "params": {"token_reduce_to": 5000}}
    """

    type: Literal["middle_out"]
    params: MiddleOutParams


# Union type for all edit strategies
# When adding new strategies, add them to this Union: EditStrategy = Union[RemoveToolResultStrategy, OtherStrategy, ...]
EditStrategy = Union[
    RemoveToolResultStrategy,
    RemoveToolCallParamsStrategy,
    TokenLimitStrategy,
    MiddleOutStrategy,
]


class Asset(BaseModel):
    """Asset model representing a file asset."""

    bucket: str = Field(..., description="S3 bucket name")
    s3_key: str = Field(..., description="S3 key")
    etag: str = Field(..., description="ETag")
    sha256: str = Field(..., description="SHA256 hash")
    mime: str = Field(..., description="MIME type")
    size_b: int = Field(..., description="File size in bytes")


class Part(BaseModel):
    """Message part model representing a part of a message."""

    type: str = Field(
        ...,
        description="Part type: 'text', 'image', 'audio', 'video', 'file', 'tool-call', 'tool-result', 'data', 'thinking'",
    )
    text: str | None = Field(None, description="Text content for text parts")
    asset: Asset | None = Field(None, description="Asset information for media parts")
    filename: str | None = Field(None, description="Filename for file parts")
    meta: dict[str, Any] | None = Field(None, description="Optional metadata")


class Message(BaseModel):
    """Message model representing a message in a session."""

    id: str = Field(..., description="Message UUID")
    session_id: str = Field(..., description="Session UUID")
    parent_id: str | None = Field(None, description="Parent message UUID")
    role: str = Field(..., description="Message role: 'user' or 'assistant'")
    meta: dict[str, Any] = Field(..., description="Message metadata")
    parts: list[Part] = Field(..., description="List of message parts")
    task_id: str | None = Field(None, description="Task UUID if associated with a task")
    session_task_process_status: str = Field(
        ...,
        description="Task process status: 'success', 'failed', 'running', 'pending', 'disable_tracking', or 'limit_exceed'",
    )
    created_at: str = Field(..., description="ISO 8601 formatted creation timestamp")
    updated_at: str = Field(..., description="ISO 8601 formatted update timestamp")


class Session(BaseModel):
    """Session model representing a session."""

    id: str = Field(..., description="Session UUID")
    project_id: str = Field(..., description="Project UUID")
    user_id: str | None = Field(None, description="User UUID")
    disable_task_tracking: bool = Field(
        False, description="Whether task tracking is disabled for this session"
    )
    configs: dict[str, Any] | None = Field(
        None, description="Session configuration dictionary"
    )
    created_at: str = Field(..., description="ISO 8601 formatted creation timestamp")
    updated_at: str = Field(..., description="ISO 8601 formatted update timestamp")


class TaskData(BaseModel):
    """Task data model representing the structured data stored in a task.

    This schema matches the TaskData model in acontext_core/schema/session/task.py
    and the Go API TaskData struct.
    """

    task_description: str = Field(..., description="Description of the task")
    progresses: list[str] | None = Field(
        None, description="List of progress updates for the task"
    )
    user_preferences: list[str] | None = Field(
        None, description="List of user preferences related to the task"
    )


class Task(BaseModel):
    """Task model representing a task in a session."""

    id: str = Field(..., description="Task UUID")
    session_id: str = Field(..., description="Session UUID")
    project_id: str = Field(..., description="Project UUID")
    order: int = Field(..., description="Task order")
    data: TaskData = Field(..., description="Structured task data")
    status: str = Field(
        ...,
        description="Task status: 'success', 'failed', 'running', or 'pending'",
    )
    is_planning: bool = Field(..., description="Whether the task is in planning phase")
    created_at: str = Field(..., description="ISO 8601 formatted creation timestamp")
    updated_at: str = Field(..., description="ISO 8601 formatted update timestamp")


class ListSessionsOutput(BaseModel):
    """Response model for listing sessions."""

    items: list[Session] = Field(..., description="List of sessions")
    next_cursor: str | None = Field(None, description="Cursor for pagination")
    has_more: bool = Field(..., description="Whether there are more items")


class PublicURL(BaseModel):
    """Public URL model for asset presigned URLs."""

    url: str = Field(..., description="Presigned URL")
    expire_at: str = Field(..., description="Expiration time in ISO 8601 format")


class SessionEvent(BaseModel):
    """Session event model representing an event in a session."""

    id: str = Field(..., description="Event UUID")
    session_id: str = Field(..., description="Session UUID")
    project_id: str = Field(..., description="Project UUID")
    type: str = Field(..., description="Event type discriminator")
    data: dict[str, Any] = Field(..., description="Event payload")
    created_at: str = Field(..., description="ISO 8601 formatted creation timestamp")
    updated_at: str = Field(..., description="ISO 8601 formatted update timestamp")


class GetMessagesOutput(BaseModel):
    """Response model for getting messages.

    Note: The items field type depends on the format parameter:
    - format="acontext": items is list[Message] (Acontext format)
    - format="openai": items is OpenAI format messages
    - format="anthropic": items is Anthropic format messages

    Since format is a runtime parameter, items uses list[Any] for flexibility.
    """

    items: list[Any] = Field(
        ...,
        description="List of messages in the requested format (Message, OpenAI format, or Anthropic format)",
    )
    ids: list[str] = Field(
        ...,
        description="List of message UUIDs corresponding to each item in the same order",
    )
    metas: list[dict[str, Any]] = Field(
        default_factory=list,
        description="List of user-provided metadata for each message (same order as items/ids)",
    )
    next_cursor: str | None = Field(None, description="Cursor for pagination")
    has_more: bool = Field(..., description="Whether there are more items")
    this_time_tokens: int = Field(
        ...,
        description="Total token count of the returned messages",
    )
    public_urls: dict[str, PublicURL] | None = Field(
        None,
        description="Map of SHA256 hash to PublicURL (only included when format='acontext')",
    )
    edit_at_message_id: str | None = Field(
        None,
        description=(
            "The message ID where edit strategies were applied up to. "
            "If pin_editing_strategies_at_message was provided, this equals that value. "
            "Otherwise, this is the ID of the last message in the response. "
            "Use this value to maintain prompt cache stability by passing it as "
            "pin_editing_strategies_at_message in subsequent requests."
        ),
    )
    events: list[SessionEvent] | None = Field(
        None,
        description="Session events within the messages time window (only when with_events=True)",
    )


class GetTasksOutput(BaseModel):
    """Response model for getting tasks."""

    items: list[Task] = Field(..., description="List of tasks")
    next_cursor: str | None = Field(None, description="Cursor for pagination")
    has_more: bool = Field(..., description="Whether there are more items")


class TokenCounts(BaseModel):
    """Response model for token counts."""

    total_tokens: int = Field(
        ...,
        description="Total token count for all text and tool-call parts in a session",
    )


class MessageObservingStatus(BaseModel):
    """Response model for message observing status."""

    observed: int = Field(..., description="Number of messages with observed status")
    in_process: int = Field(
        ..., description="Number of messages with in_process status"
    )
    pending: int = Field(..., description="Number of messages with pending status")
    updated_at: str = Field(..., description="Timestamp when the status was retrieved")


class CopySessionResult(BaseModel):
    """Response model for copying a session."""

    old_session_id: str = Field(..., description="UUID of the original session")
    new_session_id: str = Field(..., description="UUID of the copied (new) session")


class ListEventsOutput(BaseModel):
    """Response model for listing session events."""

    items: list[SessionEvent] = Field(..., description="List of events")
    next_cursor: str | None = Field(None, description="Cursor for pagination")
    has_more: bool = Field(..., description="Whether there are more items")

File: src/client/acontext-ts/src/types/session.ts
/**
 * Type definitions for session, message, and task resources.
 */

import { z } from 'zod';

export const AssetSchema = z.object({
  bucket: z.string(),
  s3_key: z.string(),
  etag: z.string(),
  sha256: z.string(),
  mime: z.string(),
  size_b: z.number(),
});

export type Asset = z.infer<typeof AssetSchema>;

/**
 * Message part schema.
 * Part type is one of: 'text', 'image', 'audio', 'video', 'file', 'tool-call', 'tool-result', 'data', 'thinking'.
 */
export const PartSchema = z.object({
  type: z.string(),
  text: z.string().nullable().optional(),
  asset: AssetSchema.nullable().optional(),
  filename: z.string().nullable().optional(),
  meta: z.record(z.string(), z.unknown()).nullable().optional(),
});

export type Part = z.infer<typeof PartSchema>;

export const MessageSchema = z.object({
  id: z.string(),
  session_id: z.string(),
  parent_id: z.string().nullable(),
  role: z.string(),
  meta: z.record(z.string(), z.unknown()),
  parts: z.array(PartSchema),
  task_id: z.string().nullable(),
  /** Task process status: 'success' | 'failed' | 'running' | 'pending' | 'disable_tracking' | 'limit_exceed' */
  session_task_process_status: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Message = z.infer<typeof MessageSchema>;

export const SessionSchema = z.object({
  id: z.string(),
  project_id: z.string(),
  user_id: z.string().nullable().optional(),
  disable_task_tracking: z.boolean(),
  configs: z.record(z.string(), z.unknown()).nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Session = z.infer<typeof SessionSchema>;

/**
 * TaskData represents the structured data stored in a task.
 * This schema matches the TaskData model in acontext_core/schema/session/task.py
 * and the Go API TaskData struct.
 */
export const TaskDataSchema = z.object({
  task_description: z.string(),
  progresses: z.array(z.string()).nullable().optional(),
  user_preferences: z.array(z.string()).nullable().optional(),
});

export type TaskData = z.infer<typeof TaskDataSchema>;

export const TaskSchema = z.object({
  id: z.string(),
  session_id: z.string(),
  project_id: z.string(),
  order: z.number(),
  data: TaskDataSchema,
  status: z.string(),
  is_planning: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Task = z.infer<typeof TaskSchema>;

export const ListSessionsOutputSchema = z.object({
  items: z.array(SessionSchema),
  next_cursor: z.string().nullable().optional(),
  has_more: z.boolean(),
});

export type ListSessionsOutput = z.infer<typeof ListSessionsOutputSchema>;

export const PublicURLSchema = z.object({
  url: z.string(),
  expire_at: z.string(),
});

export type PublicURL = z.infer<typeof PublicURLSchema>;

export const SessionEventSchema = z.object({
  id: z.string(),
  session_id: z.string(),
  project_id: z.string(),
  type: z.string(),
  data: z.record(z.string(), z.unknown()),
  created_at: z.string(),
  updated_at: z.string(),
});

export type SessionEvent = z.infer<typeof SessionEventSchema>;

export const ListEventsOutputSchema = z.object({
  items: z.array(SessionEventSchema),
  next_cursor: z.string().nullable().optional(),
  has_more: z.boolean(),
});

export type ListEventsOutput = z.infer<typeof ListEventsOutputSchema>;

export const GetMessagesOutputSchema = z.object({
  items: z.array(z.unknown()),
  ids: z.array(z.string()),
  /** User-provided metadata for each message (same order as items/ids) */
  metas: z.array(z.record(z.string(), z.unknown())).default([]),
  next_cursor: z.string().nullable().optional(),
  has_more: z.boolean(),
  /** Total token count of the returned messages */
  this_time_tokens: z.number(),
  public_urls: z.record(z.string(), PublicURLSchema).nullable().optional(),
  /**
   * The message ID where edit strategies were applied up to.
   * If pin_editing_strategies_at_message was provided, this equals that value.
   * Otherwise, this is the ID of the last message in the response.
   * Use this value to maintain prompt cache stability by passing it as
   * pin_editing_strategies_at_message in subsequent requests.
   */
  edit_at_message_id: z.string().nullable().optional(),
  /** Session events within the messages time window (only when withEvents is true) */
  events: z.array(SessionEventSchema).nullable().optional(),
});

export type GetMessagesOutput = z.infer<typeof GetMessagesOutputSchema>;

export const GetTasksOutputSchema = z.object({
  items: z.array(TaskSchema),
  next_cursor: z.string().nullable().optional(),
  has_more: z.boolean(),
});

export type GetTasksOutput = z.infer<typeof GetTasksOutputSchema>;

export const TokenCountsSchema = z.object({
  total_tokens: z.number(),
});

export type TokenCounts = z.infer<typeof TokenCountsSchema>;

export const MessageObservingStatusSchema = z.object({
  observed: z.number(),
  in_process: z.number(),
  pending: z.number(),
  updated_at: z.string(),
});

export type MessageObservingStatus = z.infer<typeof MessageObservingStatusSchema>;

export const CopySessionResultSchema = z.object({
  old_session_id: z.string(),
  new_session_id: z.string(),
});

export type CopySessionResult = z.infer<typeof CopySessionResultSchema>;

/**
 * Parameters for the remove_tool_result edit strategy.
 */
export const RemoveToolResultParamsSchema = z.object({
  /**
   * Number of most recent tool results to keep with original content.
   * @default 3
   */
  keep_recent_n_tool_results: z.number().optional(),

  /**
   * Custom text to replace old tool results with.
   * @default "Done"
   */
  tool_result_placeholder: z.string().optional(),

  /**
   * List of tool names that should never have their results removed.
   * Tool results from these tools are always kept regardless of keep_recent_n_tool_results.
   */
  keep_tools: z.array(z.string()).optional(),

  /**
   * Only remove tool results whose text has more than this many tokens.
   * If omitted, all tool results are eligible for removal.
   */
  gt_token: z.number().int().min(1).optional(),
});

export type RemoveToolResultParams = z.infer<typeof RemoveToolResultParamsSchema>;

/**
 * Parameters for the remove_tool_call_params edit strategy.
 */
export const RemoveToolCallParamsParamsSchema = z.object({
  /**
   * Number of most recent tool calls to keep with full parameters.
   * @default 3
   */
  keep_recent_n_tool_calls: z.number().optional(),

  /**
   * List of tool names that should never have their parameters removed.
   * Tool calls for these tools always keep their full parameters regardless of keep_recent_n_tool_calls.
   */
  keep_tools: z.array(z.string()).optional(),

  /**
   * Only remove tool call params whose arguments have more than this many tokens.
   * If omitted, all tool calls are eligible for removal.
   */
  gt_token: z.number().int().min(1).optional(),
});
export type RemoveToolCallParamsParams = z.infer<typeof RemoveToolCallParamsParamsSchema>;

/**
 * Edit strategy to remove parameters from old tool-call parts.
 * 
 * Keeps the most recent N tool calls with full parameters, replacing older
 * tool call arguments with empty JSON "{}". The tool call ID and name remain
 * intact so tool-results can still reference them.
 * 
 * Example: { type: 'remove_tool_call_params', params: { keep_recent_n_tool_calls: 5 } }
 */
export const RemoveToolCallParamsStrategySchema = z.object({
  type: z.literal('remove_tool_call_params'),
  params: RemoveToolCallParamsParamsSchema,
});
export type RemoveToolCallParamsStrategy = z.infer<typeof RemoveToolCallParamsStrategySchema>;

/**
 * Edit strategy to replace old tool results with placeholder text.
 * 
 * Example: { type: 'remove_tool_result', params: { keep_recent_n_tool_results: 5, tool_result_placeholder: 'Cleared' } }
 */
export const RemoveToolResultStrategySchema = z.object({
  type: z.literal('remove_tool_result'),
  params: RemoveToolResultParamsSchema,
});

export type RemoveToolResultStrategy = z.infer<typeof RemoveToolResultStrategySchema>;

/**
 * Parameters for the token_limit edit strategy.
 */
export const TokenLimitParamsSchema = z.object({
  /**
   * Maximum number of tokens to keep. Required parameter.
   * Messages will be removed from oldest to newest until total tokens <= limit_tokens.
   * Tool-call and tool-result pairs are always removed together.
   */
  limit_tokens: z.number(),
});

export type TokenLimitParams = z.infer<typeof TokenLimitParamsSchema>;

/**
 * Edit strategy to truncate messages based on token count.
 * 
 * Removes oldest messages until the total token count is within the specified limit.
 * Maintains tool-call/tool-result pairing - when removing a message with tool-calls,
 * the corresponding tool-result messages are also removed.
 * 
 * Example: { type: 'token_limit', params: { limit_tokens: 20000 } }
 */
export const TokenLimitStrategySchema = z.object({
  type: z.literal('token_limit'),
  params: TokenLimitParamsSchema,
});

export type TokenLimitStrategy = z.infer<typeof TokenLimitStrategySchema>;

/**
 * Parameters for the middle_out edit strategy.
 */
export const MiddleOutParamsSchema = z.object({
  token_reduce_to: z.number(),
});

export type MiddleOutParams = z.infer<typeof MiddleOutParamsSchema>;

/**
 * Edit strategy to reduce prompt size by removing middle messages.
 *
 * Example: { type: 'middle_out', params: { token_reduce_to: 5000 } }
 */
export const MiddleOutStrategySchema = z.object({
  type: z.literal('middle_out'),
  params: MiddleOutParamsSchema,
});

export type MiddleOutStrategy = z.infer<typeof MiddleOutStrategySchema>;

/**
 * Union schema for all edit strategies.
 * When adding new strategies, extend this union: z.union([RemoveToolResultStrategySchema, OtherStrategySchema, ...])
 */
export const EditStrategySchema = z.union([
  RemoveToolResultStrategySchema,
  RemoveToolCallParamsStrategySchema,
  TokenLimitStrategySchema,
  MiddleOutStrategySchema,
]);

export type EditStrategy = z.infer<typeof EditStrategySchema>;
